### PR TITLE
feat: add index for stacks block id in nakamoto_block_headers

### PR DESCRIPTION
**Pull request by kantai**
### Description

Optimization: add index for `index_block_hash` in `nakamoto_block_headers`

### CODE REVIEW

**Review by hstove — DISMISSED** [review]

**Review by saralab — APPROVED** [review]

**Review by jcnelson — DISMISSED** [review]
Do we need a new schema version?

**Comment by jcnelson** [comment]
Ah, yes, you're right -- specifically for the chainstate (we really should do something about that too -- I bet those transactions are really unecessary)

**Comment by kantai** [comment]
> Ah, yes, you're right -- specifically for the chainstate (we really should do something about that too -- I bet those transactions are really unecessary)

Nope, I was wrong!

```rust
        if create_flag {
            // instantiate!
            StacksChainState::instantiate_db(mainnet, chain_id, index_path, true)
        } else {
            let mut marf = StacksChainState::open_index(index_path)?;
            if !Self::need_schema_migrations(marf.sqlite_conn(), mainnet, chain_id)? {
                return Ok(marf);
            }

            // need a migration
            let tx = marf.storage_tx()?;
            StacksChainState::apply_schema_migrations(&tx, mainnet, chain_id)?;
            StacksChainState::add_indexes(&tx)?;
            tx.commit()?;
```

Will add a schema version.

File: stackslib/src/chainstate/stacks/db/mod.rs [hunk 1]
```diff
@@ -1132,6 +1133,15 @@ impl StacksChainState {
                         tx.execute_batch(cmd)?;
                     }
                 }
+                "7" => {
+                    // migrate to nakamoto 3
```
Comment (CharlieC3) [inline stackslib/src/chainstate/stacks/db/mod.rs:1137]: Nit

```suggestion
                    // migrate to nakamoto 5
```

**Review by CharlieC3 — DISMISSED** [review]
Minor nit, but fine to ignore
:shipit:

**Review by jcnelson — DISMISSED** [review]

File: stackslib/src/chainstate/stacks/db/mod.rs [hunk 2]
```diff
@@ -299,14 +299,14 @@ impl DBConfig {
...
             StacksEpochId::Epoch10 => true,
-            StacksEpochId::Epoch20 => version_u32 >= 1 && version_u32 <= 7,
-            StacksEpochId::Epoch2_05 => version_u32 >= 2 && version_u32 <= 7,
-            StacksEpochId::Epoch21 => version_u32 >= 3 && version_u32 <= 7,
-            StacksEpochId::Epoch22 => version_u32 >= 3 && version_u32 <= 7,
-            StacksEpochId::Epoch23 => version_u32 >= 3 && version_u32 <= 7,
-            StacksEpochId::Epoch24 => version_u32 >= 3 && version_u32 <= 7,
-            StacksEpochId::Epoch25 => version_u32 >= 3 && version_u32 <= 7,
-            StacksEpochId::Epoch30 => version_u32 >= 3 && version_u32 <= 7,
+            StacksEpochId::Epoch20 => version_u32 >= 1 && version_u32 <= 8,
```
Comment (jferrant) [inline stackslib/src/chainstate/stacks/db/mod.rs:302]: This is a nit. Should this be set to instead of hard coded 8, "CHAINSTATE_VERSION" ?

**Review by jferrant — APPROVED** [review]

File: stackslib/src/chainstate/stacks/db/mod.rs (hunk 1 above)
Comment (obycode) [reply stackslib/src/chainstate/stacks/db/mod.rs:1136]: While you're at it, may as well change the one above to 4 as well.

**Review by obycode — DISMISSED** [review]
LGTM, just the comment nit.

**Review by jferrant — APPROVED** [review]

**Review by CharlieC3 — DISMISSED** [review]

File: stackslib/src/chainstate/stacks/db/mod.rs (hunk 2 above)
Comment (kantai) [reply stackslib/src/chainstate/stacks/db/mod.rs:302]: The idea with the code structured as it currently is, is that when the chainstate schema version changes, they need to make sure that it actually supports the old epochs. I am not a huge fan of this particular switch statement (I don't think its very good at describing the actual compatibility), but I'm loathe to upset the apple cart too much here while I'm making other changes. I'd prefer to do that in a separate PR.

File: stackslib/src/chainstate/stacks/db/mod.rs (hunk 1 above)
Comment (kantai) [reply stackslib/src/chainstate/stacks/db/mod.rs]: I just removed the comments -- at best, they're just restating the log lines beneath them.

**Review by CharlieC3 — APPROVED** [review]
LGTM!

**Review by obycode — APPROVED** [review]

**Pull request merged**